### PR TITLE
feat (gocardless): handle gocardless customers

### PR DESCRIPTION
**Pull request by lovrocolic**
## Context

This PR handles gocardless customers. 

## Description

We follow the same logic as for Stripe here. The only difference is that property that determines if customer will be synced in provider or not is moved from organisation level to customer lavel.

### CODE REVIEW

File: spec/services/payment_provider_customers/gocardless_service_spec.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        gocardless_service.create
+
+        expect(gocardless_customers_service).not_to have_received(:create)
+      end
+    end
+
+    context 'when failing to create the customer' do
+      it 'delivers an error webhook' do
+        allow(GoCardlessPro::Client).to receive(:new)
+          .and_raise(GoCardlessPro::ApiError.new({'message' => 'error'}))
```
Comment (jdenquin) [inline spec/services/payment_provider_customers/gocardless_service_spec.rb:58]: Layout/SpaceInsideHashLiteralBraces: Space inside } missing. (https://rubystyle.guide#spaces-braces)

File: app/services/customers/create_service.rb [hunk 2]
```diff
@@ -78,47 +79,55 @@ def create(**args)
...
     end
 
     def create_or_update_provider_customer(customer, billing_configuration = {})
-      create_result = PaymentProviderCustomers::CreateService.new(customer).create_or_update(
-        customer_class: PaymentProviderCustomers::StripeCustomer,
-        payment_provider_id: customer.organization.stripe_payment_provider&.id,
-        params: billing_configuration,
-        async: !(billing_configuration || {})[:sync],
-      )
+      create_result = if billing_configuration[:payment_provider] == 'stripe'
```
Comment (jdenquin) [inline app/services/customers/create_service.rb:115]: Style/CaseLikeIf: Convert `if-elsif` to `case-when`. (https://rubystyle.guide#case-vs-if-else)

File: app/services/customers/update_service.rb [hunk 3]
```diff
@@ -70,12 +72,26 @@ def update_currency(customer:, currency:, customer_update: false)
...
-    def create_or_update_provider_customer(customer, billing_configuration = {})
-      handle_stripe_customer = customer.payment_provider.present?
-      handle_stripe_customer ||= (billing_configuration || {})[:provider_customer_id].present?
-      handle_stripe_customer ||= customer.stripe_customer&.provider_customer_id.present?
-      return unless handle_stripe_customer
+    def create_or_update_provider_customer(customer, payment_provider, billing_configuration = {})
+      handle_provider_customer = customer.payment_provider.present?
+      handle_provider_customer ||= (billing_configuration || {})[:provider_customer_id].present?
+
+      if payment_provider == 'stripe'
```
Comment (jdenquin) [inline app/services/customers/update_service.rb:79]: Style/CaseLikeIf: Convert `if-elsif` to `case-when`. (https://rubystyle.guide#case-vs-if-else)

File: app/services/customers/create_service.rb [hunk 4]
```diff
@@ -78,47 +79,56 @@ def create(**args)
...
+
+      create_or_update_provider_customer(customer, billing_configuration)
     end
 
     def handle_api_billing_configuration(customer, params, new_customer)
-      unless params.key?(:billing_configuration)
-        create_billing_configuration(customer) if new_customer
-        return
-      end
+      return unless params.key?(:billing_configuration)
```
Comment (vincent-pochet) [inline app/services/customers/create_service.rb:95]: With this change we will not create the stripe customer automatically if the organization option is checked. Is it intended?

File: app/services/customers/create_service.rb [hunk 5]
```diff
@@ -78,47 +79,56 @@ def create(**args)
...
       billing_configuration = params[:billing_configuration]
 
-      unless billing_configuration[:payment_provider] == 'stripe'
+      if new_customer
+        create_billing_configuration(customer, billing_configuration, is_api: true)
+        return
+      end
+
+      unless billing_configuration[:payment_provider] == 'stripe' ||
+             billing_configuration[:payment_provider] == 'gocardless'
```
Comment (vincent-pochet) [inline app/services/customers/create_service.rb:105]:
```suggestion
      unless %w[stripe gocardless].include?(billing_configuration[:payment_provider])
```

File: app/services/customers/create_service.rb [hunk 6]
```diff
@@ -78,47 +79,56 @@ def create(**args)
...
+                          async: !(billing_configuration || {})[:sync],
+                        )
+                      when 'gocardless'
+                        PaymentProviderCustomers::CreateService.new(customer).create_or_update(
+                          customer_class: PaymentProviderCustomers::GocardlessCustomer,
+                          payment_provider_id: customer.organization.gocardless_payment_provider&.id,
+                          params: billing_configuration,
+                          async: !(billing_configuration || {})[:sync],
+                        )
+      end
```
Comment (vincent-pochet) [inline app/services/customers/create_service.rb:130]: I'm proposing something like this to make it a bit shorter:
```suggestion
      provider_class = case billing_configuration[:payment_provider]
                                 when 'stripe'
                                   PaymentProviderCustomers::StripeCustomer
                                 when 'gocardless'
                                   PaymentProviderCustomers::GocardlessCustomer
                                 end
                                 
      create_result = PaymentProviderCustomers::CreateService.new(customer).create_or_update(
        customer_class: provider_class,
        payment_provider_id: customer.organization.payment_provider(billing_configuration[:payment_provider])&.id,
        params: billing_configuration,
        async: !(billing_configuration || {})[:sync],
      )
```

The `provider_class` initialization could even be performed in a different method

File: app/services/payment_provider_customers/gocardless_service.rb [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+      @organization ||= customer.organization
+    end
+
+    def access_token
+      organization.gocardless_payment_provider.access_token
+    end
+
+    def create_gocardless_customer
+      GoCardlessPro::Client
+        .new(access_token: access_token, environment: :sandbox)
```
Comment (vincent-pochet) [inline app/services/payment_provider_customers/gocardless_service.rb:43]: Do we want to keep this `:sandbox` config?

File: app/controllers/api/v1/customers_controller.rb [hunk 8]
```diff
@@ -97,6 +97,8 @@ def create_params
             :payment_provider,
             :provider_customer_id,
             :sync,
+            :gocardless_mandate_id,
```
Comment (vincent-pochet) [inline app/controllers/api/v1/customers_controller.rb:100]: Just to make sure I understand well, for gocardless we will have the two identifier `provider_customer_id` and `gocardless_mandate_id`?

File: app/graphql/types/payment_provider_customers/provider_input.rb [hunk 9]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+module Types
+  module PaymentProviderCustomers
+    class ProviderInput < BaseInputObject
+      graphql_name 'ProviderCustomerInput'
+
+      argument :provider_customer_id, ID, required: false
+      argument :sync_with_provider, Boolean, required: false
+      argument :gocardless_mandate_id, String, required: false
```
Comment (vincent-pochet) [inline app/graphql/types/payment_provider_customers/provider_input.rb:10]: Same comment for a `provider_` prefix

**Review by vincent-pochet — COMMENTED** [review]
Few comments but looks good! :rocket:

File: app/graphql/types/payment_provider_customers/gocardless.rb [hunk 10]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+module Types
+  module PaymentProviderCustomers
+    class Gocardless < Types::BaseObject
+      graphql_name 'GocardlessCustomer'
+
+      field :id, ID, null: false
+      field :provider_customer_id, ID, null: true
+      field :mandate_id, String, null: true
```
[the remark this answers is not in the document]
Comment (lovrocolic) [reply app/graphql/types/payment_provider_customers/gocardless.rb:10]: will fix it :)

File: app/services/customers/create_service.rb [hunk 11]
```diff
@@ -78,47 +79,56 @@ def create(**args)
 
     private
 
-    # NOTE: Check if a payment provider is configured in the organization and
-    #       force creation of provider customers
-    def create_billing_configuration(customer, billing_configuration = {})
-      create_stripe_customer = customer.organization.stripe_payment_provider&.create_customers
-      create_stripe_customer ||= (billing_configuration || {})[:provider_customer_id]
-      return unless create_stripe_customer
+    def create_billing_configuration(customer, billing_configuration = {}, is_api:)
```
[the remark this answers is not in the document]
Comment (lovrocolic) [reply app/services/customers/create_service.rb:82]: Will be done in another PR:)

File: app/services/customers/create_service.rb (hunk 4 above)
Comment (lovrocolic) [reply app/services/customers/create_service.rb:95]: exactly, we only check property on the customer level

File: app/controllers/api/v1/customers_controller.rb (hunk 8 above)
Comment (lovrocolic) [reply app/controllers/api/v1/customers_controller.rb:100]: yep

File: app/graphql/types/payment_provider_customers/provider_input.rb (hunk 9 above)
Comment (lovrocolic) [reply app/graphql/types/payment_provider_customers/provider_input.rb:10]: will fix it in another PR

**Comment by lovrocolic** [comment]
After the talk with Vincent, we agreed to merge this one and fix 'context' comment and 'mandate_id' renaming in another PR

**Pull request merged**